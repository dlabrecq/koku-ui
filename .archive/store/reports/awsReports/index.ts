import * as awsReportsActions from './awsReportsActions';
import { awsReportsStateKey } from './awsReportsCommon';
import {
  AwsCachedReport,
  AwsReportsAction,
  awsReportsReducer,
  AwsReportsState,
} from './awsReportsReducer';
import * as awsReportsSelectors from './awsReportsSelectors';

export {
  AwsReportsAction,
  AwsCachedReport,
  awsReportsActions,
  awsReportsReducer,
  awsReportsSelectors,
  AwsReportsState,
  awsReportsStateKey,
};
