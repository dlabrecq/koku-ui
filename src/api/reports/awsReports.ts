import axios from 'axios';
import { Omit } from 'react-redux';
import { isBetaFeature } from 'utils/feature';
import { getCostType } from 'utils/localStorage';

import { Report, ReportData, ReportItem, ReportItemValue, ReportMeta, ReportType, ReportValue } from './report';

export interface AwsReportItem extends ReportItem {
  account?: string;
  account_alias?: string;
  instance_type?: string;
  org_unit_id?: string;
  region?: string;
  service?: string;
}

export interface GroupByAccountData extends Omit<AwsReportData, 'accounts'> {
  account: string;
}

export interface GroupByRegionData extends Omit<AwsReportData, 'regions'> {
  region: string;
}

export interface GroupByServiceData extends Omit<AwsReportData, 'services'> {
  service: string;
}

export interface GroupByInstanceTypeData extends Omit<AwsReportData, 'instance_types'> {
  instance_type: string;
}

export interface AwsReportData extends ReportData {
  accounts?: GroupByAccountData[];
  instance_types?: GroupByInstanceTypeData[];
  regions?: GroupByRegionData[];
  services?: GroupByServiceData[];
}

export interface AwsReportMeta extends ReportMeta {
  total?: {
    cost?: ReportItemValue;
    infrastructure?: ReportItemValue;
    supplementary?: ReportItemValue;
    usage?: ReportValue;
  };
}

export interface AwsReport extends Report {
  meta: AwsReportMeta;
  data: AwsReportData[];
}

export const ReportTypePaths: Partial<Record<ReportType, string>> = {
  [ReportType.cost]: 'reports/aws/costs/',
  [ReportType.database]: 'reports/aws/costs/',
  [ReportType.network]: 'reports/aws/costs/',
  [ReportType.storage]: 'reports/aws/storage/',
  [ReportType.instanceType]: 'reports/aws/instance-types/',
};

export function runReport(reportType: ReportType, query: string) {
  const path = ReportTypePaths[reportType];

  // Todo: Show in-progress features in beta environment only
  if (isBetaFeature()) {
    switch (reportType) {
      case ReportType.cost:
      case ReportType.database:
      case ReportType.network:
        return axios.get<AwsReport>(`${path}?cost_type=${getCostType()}&${query}`);
    }
  }
  return axios.get<AwsReport>(`${path}?${query}`);
}
