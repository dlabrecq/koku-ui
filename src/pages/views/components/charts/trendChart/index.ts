export { TrendChart, TrendChartProps } from './trendChart';
