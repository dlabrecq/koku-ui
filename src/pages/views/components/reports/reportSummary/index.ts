export { ReportSummary } from './reportSummary';
export { ReportSummaryAlt } from './reportSummaryAlt';
export { ReportSummaryDetails } from './reportSummaryDetails';
export { ReportSummaryItem } from './reportSummaryItem';
export { ReportSummaryItems } from './reportSummaryItems';
export { ReportSummaryCost } from './reportSummaryCost';
export { ReportSummaryDailyCost } from './reportSummaryDailyCost';
export { ReportSummaryDailyTrend } from './reportSummaryDailyTrend';
export { ReportSummaryTrend } from './reportSummaryTrend';
export { ReportSummaryUsage } from './reportSummaryUsage';
