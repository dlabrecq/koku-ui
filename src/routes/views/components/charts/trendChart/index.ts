export { default as TrendChart, TrendChartProps } from './trendChart';
