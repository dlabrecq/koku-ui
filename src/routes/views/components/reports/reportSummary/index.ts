export { default as ReportSummary, ReportSummaryProps } from './reportSummary';
export { default as ReportSummaryAlt, ReportSummaryAltAltProps } from './reportSummaryAlt';
export { default as ReportSummaryDetails, ReportSummaryDetailsProps } from './reportSummaryDetails';
export { default as ReportSummaryItem, ReportSummaryItemProps } from './reportSummaryItem';
export { default as ReportSummaryItems, ReportSummaryItemsProps } from './reportSummaryItems';
export { default as ReportSummaryCost, ReportSummaryCostProps } from './reportSummaryCost';
export { default as ReportSummaryDailyCost, ReportSummaryDailyCostProps } from './reportSummaryDailyCost';
export { default as ReportSummaryDailyTrend, ReportSummaryDailyTrendProps } from './reportSummaryDailyTrend';
export { default as ReportSummaryTrend, ReportSummaryTrendProps } from './reportSummaryTrend';
export { default as ReportSummaryUsage, ReportSummaryUsageProps } from './reportSummaryUsage';
