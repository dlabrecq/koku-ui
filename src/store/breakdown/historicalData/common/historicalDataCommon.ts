import type { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum HistoricalDataWidgetType {
  cost = 'cost', // This type displays historical cost chart
  network = 'network', // This type displays historical network chart
  trend = 'trend', // This type displays historical trend chart
  usage = 'usage', // This type displays historical usage chart
  volume = 'volume', // This type displays historical volume chart
}

export interface HistoricalDataWidget {
  chartName: string; // Will be the prefix for ids within the chart
  id: number;
  network?: {
    showWidgetOnGroupBy?: string[]; // Show network chart when group_by is matched
  };
  reportPathsType: ReportPathsType; // Report URL path
  reportType: ReportType; // Report type; cost, storage, etc.
  type: HistoricalDataWidgetType;
}
