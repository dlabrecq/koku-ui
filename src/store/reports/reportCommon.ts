import { ReportPathsType, ReportType } from 'api/reports/report';
import { isBetaFeature } from 'utils/feature';
import { getCostType } from 'utils/localStorage';
export const reportStateKey = 'report';

export function getReportId(reportPathsType: ReportPathsType, reportType: ReportType, query: string) {
  // Todo: Show in-progress features in beta environment only
  if (isBetaFeature()) {
    switch (reportType) {
      case ReportType.cost:
      case ReportType.database:
      case ReportType.network:
        return `${reportPathsType}--${reportType}--${getCostType()}--${query}`;
    }
  }
  return `${reportPathsType}--${reportType}--${query}`;
}
