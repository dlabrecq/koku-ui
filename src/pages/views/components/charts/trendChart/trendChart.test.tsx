jest.mock('date-fns/format');

import { Chart, ChartArea } from '@patternfly/react-charts';
import { AwsReport, AwsReportData } from 'api/reports/awsReports';
import { shallow } from 'enzyme';
import * as utils from 'pages/views/components/charts/common/chartDatumUtils';
import React from 'react';

import { TrendChart, TrendChartProps } from './trendChart';

const currentMonthReport: AwsReport = createReport('1-15-18');
const previousMonthReport: AwsReport = createReport('12-15-17');

const currentData = utils.transformReport(currentMonthReport, utils.ChartType.daily);
const previousData = utils.transformReport(previousMonthReport, utils.ChartType.daily);

const props: TrendChartProps = {
  title: 'Trend Title',
  height: 100,
  currentData,
  previousData,
  formatter: jest.fn(),
  formatOptions: {},
};

test('reports are formatted to datums', () => {
  const view = shallow(<TrendChart {...props} />);
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(2);
  expect(charts.at(0).prop('data')).toMatchSnapshot('previous month data');
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('null previous and current reports are handled', () => {
  const view = shallow(<TrendChart {...props} currentData={null} previousData={null} />);
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(2);
});

test('height from props is used', () => {
  const view = shallow(<TrendChart {...props} />);
  expect(view.find(Chart).prop('height')).toBe(props.height);
});

test('labels formats with datum and value formatted from props', () => {
  const tooltipFormatMock = jest.spyOn(utils, 'getTooltipContent');
  const formatLabel = jest.fn();
  tooltipFormatMock.mockImplementation(() => formatLabel);

  const view = shallow(<TrendChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '1-1-1',
    units: 'hrs',
  };
  const group = view.find(Chart);
  group.props().containerComponent.props.labels({ datum });
  expect(formatLabel).toBeCalledWith(datum.y, datum.units, props.formatOptions);
  expect(view.find(Chart).prop('height')).toBe(props.height);
});

test('trend is a running total', () => {
  const multiDayReport: AwsReport = {
    data: [createReportDataPoint('1-15-18', 1), createReportDataPoint('1-16-18', 2)],
  };
  const multiDaytData = utils.transformReport(multiDayReport, utils.ChartType.daily);
  const view = shallow(<TrendChart {...props} currentData={multiDaytData} />);
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('trend is a daily value', () => {
  const multiDayReport: AwsReport = {
    data: [createReportDataPoint('1-15-18', 1), createReportDataPoint('1-16-18', 2)],
  };
  const multiDaytData = utils.transformReport(multiDayReport, utils.ChartType.daily);
  const view = shallow(<TrendChart {...props} currentData={multiDaytData} />);
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

function createReport(date: string): AwsReport {
  return {
    data: [createReportDataPoint(date)],
  };
}

function createReportDataPoint(date: string, cost = 1): AwsReportData {
  const item = { value: cost, units: 'unit' };
  return {
    date,
    values: [{ date, cost: item }],
  };
}
