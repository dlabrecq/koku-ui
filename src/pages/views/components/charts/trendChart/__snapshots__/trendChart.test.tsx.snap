// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reports are formatted to datums: current month data 1`] = `Array []`;

exports[`reports are formatted to datums: previous month data 1`] = `Array []`;

exports[`trend is a daily value: current month data 1`] = `Array []`;

exports[`trend is a running total: current month data 1`] = `Array []`;
